pkg/backends/calico/routes.go

```diff
@@ -267,21 +349,56 @@ func (rg *routeGenerator) unsetRouteForSvc(obj interface{}) {
 	defer rg.Unlock()
 
 	// Remove the route if it exists.
-	if route, exists := rg.svcRouteMap[key]; exists {
-		rg.withdrawRoute(key, route)
+	if route, exists := rg.svcClusterRouteMap[key]; exists {
+		rg.withdrawClusterRoute(key, route)
 	}
+
+	// Remove all External IP's
+	rg.unsetExternalRoutesForSvc(key)
+
 }
 
-// advertiseRoute advertises a route and caches it.
-func (rg *routeGenerator) advertiseRoute(key, route string) {
-	rg.svcRouteMap[key] = route
+// unsetExternalRoutesForSvc withdraws all routes for the service with
+// the given key.
+func (rg *routeGenerator) unsetExternalRoutesForSvc(key string) {
+	if rg.svcExternalRouteMap[key] != nil {
+		for route := range rg.svcExternalRouteMap[key] {
+			rg.withdrawExternalRoute(key, route)
+		}
+	}
+}
+
+// advertiseClusterRoute advertises a route for a service ClusterIP and caches it.
+func (rg *routeGenerator) advertiseClusterRoute(key, route string) {
+	rg.svcClusterRouteMap[key] = route
 	rg.client.AddStaticRoutes([]string{route})
 }
 
-// withdrawRoute withdraws a route and removes it from the cache.
-func (rg *routeGenerator) withdrawRoute(key, route string) {
+// withdrawClusterRoute withdraws a route and removes it from the ClusterIP cache.
+func (rg *routeGenerator) withdrawClusterRoute(key, route string) {
 	rg.client.DeleteStaticRoutes([]string{route})
-	delete(rg.svcRouteMap, key)
+	delete(rg.svcClusterRouteMap, key)
+}
+
+// advertiseExternalRoute advertises a route for a service ExternalIP and
+// caches it.
+func (rg *routeGenerator) advertiseExternalRoute(key, route string) {
+	if _, hasKey := rg.svcExternalRouteMap[key]; !hasKey {
+		rg.svcExternalRouteMap[key] = make(map[string]bool)
+	}
+
+	rg.client.AddStaticRoutes([]string{route})
+	rg.svcExternalRouteMap[key][route] = true
+}
+
+// withdrawExternalRoute withdraws a route for a service ExternalIP and
+// removes it from the cache.
+func (rg *routeGenerator) withdrawExternalRoute(key, route string) {
+	rg.client.DeleteStaticRoutes([]string{route})
+
+	if rg.svcExternalRouteMap[key] != nil {
+		delete(rg.svcExternalRouteMap[key], route)
+	}
```

Comment:
I think we'll want to check if `rg.svcExternalRouteMap[key]` is empty here, and if it is delete the key from it.

Otherwise we might grow this map indefinitely, leaking memory.